File: src/Algorithms/Divide_And_Conquer/53_Maximum_Subarray.js
var maxSubArray = function(nums) {

    return divide(0, nums.length - 1);

    // *********************************
    function divide(left, right) {

        if (left > right) return -Infinity;

        const mid = left + Math.floor((right - left) / 2);

        let curSum = 0;
        let leftMaxSum = 0;
        let rightMaxSum = 0;
    
        for (let i = mid - 1; i >= left; i--) {
            curSum += nums[i];
            leftMaxSum = Math.max(leftMaxSum, curSum); 
        }

        curSum = 0;
        for (let i = mid + 1; i <= right; i++) {
            curSum += nums[i]
            rightMaxSum = Math.max(rightMaxSum, curSum);
        }


        const leftDivisionSum = divide(left, mid - 1);
        const rightDivisionSum = divide(mid + 1, right);
        const midSum = leftMaxSum + nums[mid] + rightMaxSum;

        return Math.max(leftDivisionSum, rightDivisionSum, midSum);
    }


};

File: src/Algorithms/Divide_And_Conquer/53_Maximum_Subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        return divideAndConquer(nums, 0, nums.length - 1);
    }

    public int divideAndConquer(int[] nums, int left, int right) {
        if (left > right)
            return Integer.MIN_VALUE;

        int mid = left + (right - left) / 2;
        int leftHalfMaxSum = 0;
        int rightHalfMaxSum = 0;

        int curSum = 0;
        for (int i = mid - 1; i >= left; i--) {
            curSum += nums[i];
            leftHalfMaxSum = Math.max(leftHalfMaxSum, curSum);
        }

        curSum = 0;
        for (int i = mid + 1; i <= right; i++) {
            curSum += nums[i];
            rightHalfMaxSum = Math.max(rightHalfMaxSum, curSum);
        }

        int midSubArrMaxSum = leftHalfMaxSum + nums[mid] + rightHalfMaxSum;
        int leftSubArrMaxSum = divideAndConquer(nums, left, mid - 1);
        int rightSubArrMaxSum = divideAndConquer(nums, mid + 1, right);

        return Math.max(Math.max(leftSubArrMaxSum, rightSubArrMaxSum), midSubArrMaxSum);
    }
}
